iot_config/src/region_map.rs

```diff
@@ -0,0 +1,149 @@
+use futures::stream::TryStreamExt;
+use helium_proto::{BlockchainRegionParamsV1, Message, Region};
+use hextree::{
+    compaction::EqCompactor,
+    h3ron::{FromH3Index, H3Cell},
+    HexTreeMap,
+};
+use libflate::gzip::Decoder;
+use std::{collections::HashMap, io::Read, sync::Arc};
+use tokio::sync::RwLock;
+
+pub struct RegionMap {
+    region_hextree: Arc<RwLock<HexTreeMap<Region, EqCompactor>>>,
+    params_hashmap: Arc<RwLock<HashMap<Region, BlockchainRegionParamsV1>>>,
+}
+
+impl RegionMap {
+    pub async fn new(db: impl sqlx::PgExecutor<'_> + Copy) -> Result<Self, RegionMapError> {
+        let region_hextree = build_region_tree(db).await?;
+        let params_map = build_params_map(db).await?;
+        Ok(Self {
+            region_hextree: Arc::new(RwLock::new(region_hextree)),
+            params_hashmap: Arc::new(RwLock::new(params_map)),
+        })
+    }
+
+    pub async fn get_region(&self, location: u64) -> Option<Region> {
+        self.region_hextree
+            .read()
+            .await
+            .get(H3Cell::from_h3index(location))
+            .cloned()
+    }
+
+    pub async fn get_params(&self, region: &Region) -> Option<BlockchainRegionParamsV1> {
+        self.params_hashmap.read().await.get(region).cloned()
+    }
+
+    pub async fn insert_params(&self, region: Region, params: BlockchainRegionParamsV1) {
+        _ = self.params_hashmap.write().await.insert(region, params)
+    }
+
+    pub async fn swap_tree(&self, new_map: HexTreeMap<Region, EqCompactor>) {
+        *self.region_hextree.write().await = new_map
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum RegionMapError {
+    #[error("region map database error")]
+    DbFetch(#[from] sqlx::Error),
+    #[error("params decode error")]
+    ParamsDecode(#[from] helium_proto::DecodeError),
+    #[error("indices gunzip error")]
+    IdxGunzip(#[from] std::io::Error),
+    #[error("unsupported region error: {0}")]
+    UnsupportedRegion(i32),
+}
+
+#[derive(sqlx::FromRow)]
+pub struct HexRegion {
+    pub region: i32,
+    pub params: Vec<u8>,
+    pub indexes: Vec<u8>,
+}
+
+pub async fn build_region_tree(
+    db: impl sqlx::PgExecutor<'_>,
+) -> Result<HexTreeMap<Region, EqCompactor>, RegionMapError> {
+    let mut region_tree = HexTreeMap::with_compactor(EqCompactor);
+
+    let mut regions = sqlx::query_as::<_, HexRegion>("select * from regions").fetch(db);
+
+    while let Some(region_row) = regions.try_next().await? {
+        let region = Region::from_i32(region_row.region)
```

Comment:
does the `from_i32` imply that the `Region` type is `repor(i32)`? If so, is there any safe way to reduce it to a `u8` to safe memory and improve lookup performance?

Reply:
near as i can tell from the `prost_types` that's correct (`repr(i32)`: https://docs.rs/prost-types/latest/prost_types/enum.Syntax.html).
so you're thinking something like pulling the i32 out of the database row, casting to u8 when inserting into the hextreemap and then casting _back_ to a `Region` enum variant when i retrieve it from the hextreemap in order to do the params lookup in the hashmap for that region?

Reply:
that's what I was thinking, but just use i32 unless memory use or lookup perf becomes a bottleneck.